zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -984,11 +997,51 @@ private void fireAllEvents() {
 
   protected List<String> getChildren(final String path, final boolean watch) {
     long startT = System.currentTimeMillis();
+
     try {
       List<String> children = retryUntilConnected(new Callable<List<String>>() {
+        private int connectionLossRetryCount = 0;
+
         @Override
         public List<String> call() throws Exception {
-          return getConnection().getChildren(path, watch);
+          try {
+            return getConnection().getChildren(path, watch);
+          } catch (ConnectionLossException e) {
+            ++connectionLossRetryCount;
+            // Allow retrying 3 times before checking stat checking number of children,
+            // because there is a higher possibility that connection loss is caused by other
+            // factors such as network connectivity, connected ZK node could not serve
+            // the request, session expired, etc.
+            if (connectionLossRetryCount >= 3) {
+              // Issue: https://github.com/apache/helix/issues/962
+              // Connection loss might be caused by an excessive number of children.
+              // Infinitely retrying connecting may cause high GC in ZK server and kill ZK server.
+              // This is a workaround to check numChildren to have a chance to exit retry loop.
+              // TODO: remove this check once we have a better way to exit infinite retry
+              Stat stat = getStat(path);
+              if (stat != null) {
+                if (stat.getNumChildren() > NUM_CHILDREN_LIMIT) {
+                  LOG.error("Failed to get children for path {} because number of children {} "
+                          + "exceeds limit {}, aborting retry.", path, stat.getNumChildren(),
+                      NUM_CHILDREN_LIMIT);
+                  // There is not an accurate KeeperException for the purpose.
+                  // MarshallingErrorException could represent transport error,
+                  // so use it to exit retry loop and tell that zk is not able to
+                  // transport the data because packet length is too large.
+                  throw new KeeperException.MarshallingErrorException();
```

Comment:
BadArgumentsException looks better than this one.

Reply:
I considered both. `BadArgumentsException` was the one I tried to use, but realized that `MarshallingErrorException` could represent a transport error and mostly mean data exceeds jute.maxbuffer. Here is an example: https://tgockel.github.io/zookeeper-cpp/classzk_1_1marshalling__error.html#details
`An error occurred while marshalling data. The most common cause of this is exceeding the Jute buffer size – meaning the transaction was too large`
So I decided to choose MarshallingError.  BadArgumentsException for me sounds more like the API is misused, eg, invalid path that has extra slash "/path/".

Reply:
Sounds good. So just mention this point instead of saying there is no accurate exception? It seems to be accurate.
Another option is to throw an HelixException, or some new zkclient exception that we invent. Either way works, I think.

Reply:
OK. Let me just remove that line of comment to avoid confusion :)

Considering backward compatibility, if a `KeeperException` is thrown for the operation, it will be converted to `ZkException`. If user catches `ZkException` and handles it, there is extra impact, their app still runs.

If we throw a ZkException (we don't throw HelixException since we've separate zk module from helix), it will be converted to a RuntimeException, which may cause ZkClient users terminated if they don't catch RuntimeException. 

Or an option is: we change `retryUntilConnected` to also catch `ZkException`. With this change, we could throw `ZkException` with a message indicating "Large children error". I am not sure if we would like to change `retryUntilConnected` considering this is a workaround and later we remove the logic. What do you think about this option? @jiajunwang

Reply:
`If user catches ZkException and handles it, there is extra impact, their app still runs.`

Do you mean there is **no extra impact?**

Overall, I am a little bit confused. I think ZkException is what we (IOItech) convert from KeeperException to. Upper layer sees ZkException, they don't deal with KeeperException (which is from Zookeeper client).

In this case, as long as ZkClient event thread can catch this ZkException and don't terminate, that is what we want, right?

Am I missing something?

Reply:
@kaisun2000 Yeah you caught it. I meant there is **no** extra impact. Sorry for the confusion.

Yes, as I mentioned, there are two options:
1.  throw `KeeperException.MarshallingErrorException`. This will be converted by retryUntilConnected to ZkException. Benefit is we don't change retryUntilConnected.
2. throw ZkException. Benefit is we could add message in the exception. But we have to change retryUntilConnected to catch ZkException and rethrow it. Considering this is a workaround and later we remove the logic, I use `KeeperException.MarshallingErrorException` here. What do you think? @kaisun2000

Reply:
As discussed offline. Here is the basic principle inherited from IOITech
this layer, we see KeeperException,
user see ZkException,
retryUntilConnected, throw ZkException and use see ZkException. 

Here is an idea:
So maybe we can just add another type of ZkException say TooManyChildrenZkException and convert this from retryUntilConnect to this new ZkException. User of ZkClient can handle this TooManyChidrenZkException with whatever logic they want.

Reply:
Adding @jiajunwang for extra opinions.
I understand that and I also thought about it. My take is since this is only a workaround not for long term, I would like to just make it simple.
If later we would like to make it nicer because ZK gives a different error code, we could have a sub ZkException like `ZkMarshallingError` (we already have this exception defined).